Fixes #882 - Add our own upload backend, remove Flask-Uploads

miketaylr opened the pull request:
So, in order to manually verify the screenshots are working, you'll need to install some WIP add-ons. The easiest is probably Opera or Chrome: https://github.com/webcompat/webcompat-reporter-extensions/tree/master/opera (go to extensions, enable developer mode, point to that directory).

You also need to cherry-pick a WIP commit `git fetch origin && git cherry-pick e275f75f8a86f891215260434aaaf7d5872e1e8f`.

It's admittedly a mess, so feel free to ignore that entirely and we can verify screenshots when #879 is up for review.

"regular" file uploads should continue to work in the same way.

I haven't had time yet to figure out the testing story for the screenshot stuff, so that will come in follow up commits or issues.

r? @karlcow

miketaylr commented on webcompat/api/uploads.py:
```diff
@@ -42,21 +99,23 @@ def upload():
-            abort(415)
+    if 'image' in request.files and request.files['image'].filename:
+        imagedata = request.files['image']
+    elif 'screenshot' in request.form:
```
`screenshot` comes from an add-on `postMessage` https://github.com/webcompat/webcompat.com/blob/e275f75f8a86f891215260434aaaf7d5872e1e8f/webcompat/static/js/lib/bugform.js#L89-L117, if you're curious.

miketaylr commented on tests/test_uploads.py:
```diff
@@ -22,6 +21,47 @@
 
 
+# taken from https://github.com/srusskih/flask-uploads/blob/master/flaskext/uploads.py#L476 # nopep8
+class TestingFileStorage(FileStorage):
```
This change isn't very interesting, it's just copying and pasting from Flask-Upload so we can ditch the dependency and have our tests still working.

karlcow replied:
nothing forbids you to put the comment under the class. :)

karlcow commented on run.py:
```diff
@@ -31,8 +31,8 @@
-    if e.import_name == 'config':
-        # config not found, somebody forgot to make a copy of config.py.example and modify?
+    if 'import_name' in e and e.import_name == 'config':
+        # config not found, did you forget to copy config.py.example?
```
nitpick: I was about to fix this in another commit. Here the comment says what the dev should do (which is the error message), instead of documenting the code. ^_^ I would put instead. 

```python
# Catching the missing config.py file
```
The `NO_CONF_ERROR` explains what the dev has to do.

miketaylr replied:
:+1:

miketaylr commented on webcompat/views.py:
```diff
@@ -245,7 +245,7 @@ def download_file(filename):
         '''
         return send_from_directory(
-            app.config['UPLOADS_DEFAULT_DEST'] + '/uploads', filename)
+            app.config['UPLOADS_DEFAULT_DEST'], filename)
```
`/uploads` was added to `[UPLOADS_DEFAULT_DEST]`, because that's specific to how someone wants to set up their system.

karlcow replied:
did you modify `config.py.example`? I don't see in this pull request?

miketaylr replied:
Oops, will do, thanks.

karlcow commented on webcompat/api/uploads.py:
```diff
@@ -26,11 +23,71 @@
+    """
+    This exception is raised when we try to save a file that is not
+    allowed. Allowed files are defined in ALLOWED_FORMATS
+    """
```
I guess I can fix it later. but I would rather have something slightly different here. Probably something derived from TypeError. But your code will work anyway here. Just a reminder for me.

karlcow commented on webcompat/api/uploads.py:
```diff
@@ -26,11 +23,71 @@
+            # data:image/png;base64, <base64 string>
+            if imagedata.split('/')[1].split(';')[0] != 'png':
+                raise UploadNotAllowed()
+            return 'png'
```
here I don't understand. Why do we allow only png. Will not it get into the way of attach screenshots in the comment and someone uploading a jpeg.

karlcow replied:
Also Pillow provides specific way to test the right file format instead of the extension which is a ok test for validity but not a real one. 

```python
>>> from PIL import Image
>>> im = Image.open('pacifique-fuji.jpg')
>>> im.format
'JPEG'
>>> 
```

Also when the file is not correct, we get an error.

``` python
>>> im2 = Image.open('foobar.pdf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/PIL/Image.py", line 2286, in open
    % (filename if filename else fp))
IOError: cannot identify image file 'foobar.pdf'
```

miketaylr replied:
> Why do we allow only png. Will not it get into the way of the feature "attach screenshots" in the comment and someone uploading a jpeg

The base64 image from the add-on is only a PNG (for now, we could extend it in the future to allow the user to choose the image encoder, if that seems useful). This codepath is different than the "regular" image upload.

miketaylr replied:
Pillow can give us the extension from a decoded base64 image, so that's what I moved it to -- no more hard-coding. \o/

karlcow commented on webcompat/api/uploads.py:
```diff
@@ -26,11 +23,71 @@
+        '''Check that the file is allowed, then save to filesystem.'''
+        if self.file_ext not in ALLOWED_FORMATS:
+            raise UploadNotAllowed()
+
```
See above.

karlcow commented:
ok some preliminary comments done. I haven't installed it yet locally. This should take a bit of time to review. Hopefully this afternoon.

miketaylr commented:
Thanks Karl, great feedback. I'm traveling tomorrow, but maybe will work on addressing comments in the evening.

miketaylr commented:
> I haven't installed it yet locally.

One more comment, if you do install the Opera add-on (or any of them), you'll need to change `var prefix` from https://github.com/webcompat/webcompat-reporter-extensions/blob/master/opera/background.js#L5 to:

`var prefix = 'http://localhost:5000/?open=1&url=';`

Then once you have it installed, you can click the new lightbulb logo button in your browser and it'll open the local version w/ a placeholder screenshot below the form (still waiting on UI). Submitting that form should result in a new issues w/ the screenshot embedded just like an upload.

karlcow commented:
ok let's see.

```
→ git checkout -b miketest upstream/issues/882/1
Branch miketest set up to track remote branch issues/882/1 from upstream.
Switched to a new branch 'miketest'
```

ok after pointing Opera Desktop to the extension I was able to make the extension inside the chrome. The issue is that the screenshot doesn't happen at all. I guess it's working only on mobile?

OK let's see with a normal upload.

```
→ find uploads -ls
48462835        0 drwxr-xr-x    3 karl             staff                 102 22 déc 16:25 uploads
48462836        0 drwxr-xr-x    3 karl             staff                 102 22 déc 16:25 uploads/2015
48462837        0 drwxr-xr-x    3 karl             staff                 102 22 déc 16:25 uploads/2015/12
48462838        8 -rw-r--r--    1 karl             staff                1309 22 déc 16:25 uploads/2015/12/e267f789-8010-401b-931e-29147dcbb268.png
```

so far so good. 

With your current code, renamed a PDF file as `foo.png`
https://github.com/webcompat/webcompat-tests/issues/314

No errors. The log 
```
127.0.0.1 - - [22/Dec/2015 16:27:02] "GET /?open=1&url=https%3A%2F%2Fwww.mozilla.org%2Ffr%2F HTTP/1.1" 200 -
127.0.0.1 - - [22/Dec/2015 16:27:08] "GET /api/issues/labels?per_page=100 HTTP/1.1" 200 -
127.0.0.1 - - [22/Dec/2015 16:27:08] "GET /api/issues/category/new HTTP/1.1" 200 -
127.0.0.1 - - [22/Dec/2015 16:27:19] "POST /issues/new HTTP/1.1" 302 -
127.0.0.1 - - [22/Dec/2015 16:27:19] "GET /thanks/314 HTTP/1.1" 200 -
```

The upload 

```
48462835        0 drwxr-xr-x    3 karl             staff                 102 22 déc 16:25 uploads
48462836        0 drwxr-xr-x    3 karl             staff                 102 22 déc 16:25 uploads/2015
48462837        0 drwxr-xr-x    4 karl             staff                 136 22 déc 16:27 uploads/2015/12
48462914      200 -rw-r--r--    1 karl             staff               99045 22 déc 16:27 uploads/2015/12/7e952d67-4413-4c1d-851e-eefbdf325d6a.png
48462838        8 -rw-r--r--    1 karl             staff                1309 22 déc 16:25 uploads/2015/12/e267f789-8010-401b-931e-29147dcbb268.png
```

But I think the previous code had the same caveat. So it just shows that it is "working" the same as before. ^_^ for this part at least so that's good. We can start to improve from here.

karlcow commented on webcompat/api/uploads.py:
```diff
@@ -26,11 +23,71 @@
+        if self.is_filestorage:
+            return imagedata.filename.split('.')[-1].lower()
+        if self.is_base_64:
+            # We only expect base64 png screenshot uploads,
```
ah interesting.

```
127.0.0.1 - - [22/Dec/2015 16:40:15] "GET /?open=1&url=https%3A%2F%2Fwww.mozilla.org%2Ffr%2F HTTP/1.1" 200 -
127.0.0.1 - - [22/Dec/2015 16:40:17] "GET /api/issues/labels?per_page=100 HTTP/1.1" 200 -
127.0.0.1 - - [22/Dec/2015 16:40:18] "GET /api/issues/category/new HTTP/1.1" 200 -
self.is_base_64    False
self.is_filestorage True
self.file_ext           jpg
127.0.0.1 - - [22/Dec/2015 16:40:39] "POST /issues/new HTTP/1.1" 302 -
127.0.0.1 - - [22/Dec/2015 16:40:39] "GET /thanks/316 HTTP/1.1" 200 -
```

The raise `UploadNotAllowed` is never raised when `self.is_base_64  =  False` in `def get_file_ext(self, imagedata):`.

karlcow commented:
ok I start to have a better understanding of the code. Either we receive from `image` or from `screenshot`, in both cases we get `imagedata`. That could be nice to log this for seeing if we have a lot of screenshot vs image.

Then the code is going to the initialization of `Upload` under a `try:` statement.

The `init` computes

* save a local imagedata
* check if it's is_base_64
* check if it's is_filestorage
* extract the extension
  * is_filestorage based on the filename
  * is_base_64 based on the string

I guess we could replace with PIL. easily, after the 

``` python 
from PIL import Image
try:
    im = Image.open(blah)            
except IOError:
    # Not a valid format
    abort(415)
try:
    file_ext = im.format.lower()
    if file_ext not in ALLOWED_FORMATS:
        #not an allowed format
        raise TypeError("Image file format not allowed")
```

and you could return `file_ext`. That way we can test for anyone and it would reduce your code, also it would removed the need of a specific Exception.

miketaylr commented:
> The issue is that the screenshot doesn't happen at all.

Did you do this step? 

>  You also need to cherry-pick a WIP commit git fetch origin && git cherry-pick e275f75f8a86f891215260434aaaf7d5872e1e8f.

This pulls in a commit from another branch that puts the screenshot into the UI (but in a very crude fashion). Sorry, this is a huge pain in the butt with so many moving pieces. But thanks for reviewing so far!

miketaylr commented:
> But I think the previous code had the same caveat.

Heh, yeah. I'm maintaining bug compatibility here :see_no_evil: 

> That could be nice to log this for seeing if we have a lot of screenshot vs image.

That's a good point. I guess once we have logging in place (depending on #717), let's hook that up. I'll file a follow-up issue.

> I guess we could replace with PIL.

Yeah, this is cool. And then I guess it would make it very simple for you to extend this to the image optimization work in #710.

miketaylr commented:
I'm going to rebase fabc25e out before merging, so ignore that -- it just makes it simpler to test.  747e77b has the changes from Karl's feedback thus far.

@karlcow, what do you think of 747e77b? I decided to convert everything to a PIL Image object so we can use its `save` method, rather than special-case it for base64 stuff.

karlcow commented:
## About testing screenshot

> This pulls in a commit from another branch that puts the screenshot into the UI (but in a very crude fashion). Sorry, this is a huge pain in the butt with so many moving pieces. But thanks for reviewing so far!

I think I had done. But it's ok, because it doesn't really change the rest of the code. given that after screenshot and image it's basically the same code.

miketaylr commented:
> A couple of things possible.

Yeah, good point. I don't know why/how someone would be sending a base64 JPEG, but let's take the safer/more flexible option.

Thanks for your review @karlcow-sama :pray: :clap: !

miketaylr commented:
Also TODO: Add PIL to requirements.txt and write any docs about getting that set-up (I had to `brew install` a handful of things, etc.)

karlcow commented:
@miketaylr About 

>Also TODO: Add PIL to requirements.txt and write any docs about getting that set-up (I had to brew install a handful of things, etc.)

This might help you.
https://github.com/webcompat/webcompat.com/issues/710#issuecomment-147659048

miketaylr commented:
Hm, those first two commits shouldn't have been there. Looking.

miketaylr commented:
> This might help you.

Thanks Karl. I did the lazy thing and just linked to the Pillow deps install docs.

miketaylr commented:
> Hm, those first two commits shouldn't have been there. Looking.

Rebased against master, so they went away.

miketaylr commented:
OK, I've added tests both for "regular" file uploads and screenshot uploads, and fixed a bug that the tests revealed. If Travis is green, I'm gonna merge.
